File: harness/oracle/cases/message-attribution.ts
/**
 * Oracle case #18 (Wave 3 / B9): message attribution & presence.
 *
 * Drives the REAL pure helpers from v4's lib/chat/context/message-attribution.ts:
 *   filterMessagesByHistoryAccess, computePresenceWindowsForParticipant,
 *   filterMessagesByPresenceWindows, filterWhisperMessages, getParticipantName,
 *   attributeMessagesForCharacter, findUserParticipantName.
 *
 * The only impurity is `new Date(createdAt).getTime()` in the history-access
 * filter; we emit the parsed epoch-millis alongside each case so the Rust port
 * (which takes ms) uses the identical instants. The presence filters compare
 * ISO strings lexically, no parse involved.
 *
 * Run from inside the server checkout:
 *   cd ~/source/quilltap-server
 *   npx tsx ~/source/quilltap-v5/harness/oracle/cases/message-attribution.ts \
 *     > /tmp/oracle-message-attribution.ndjson
 */

import {
  filterMessagesByHistoryAccess,
  computePresenceWindowsForParticipant,
  filterMessagesByPresenceWindows,
  filterWhisperMessages,
  getParticipantName,
  attributeMessagesForCharacter,
  findUserParticipantName,
  type MessageWithParticipant,
  type PresenceWindow,
} from '@/lib/chat/context/message-attribution';
import type { ChatParticipantBase, Character } from '@/lib/schemas/types';

const asMsgs = (ms: unknown[]) => ms as unknown as MessageWithParticipant[];
const asPart = (p: unknown) => p as unknown as ChatParticipantBase;
const asParts = (ps: unknown[]) => ps as unknown as ChatParticipantBase[];
const asChars = (c: Record<string, string>): Map<string, Character> => {
  const m = new Map<string, Character>();
  for (const [cid, name] of Object.entries(c)) m.set(cid, { name } as unknown as Character);
  return m;
};
const idsOf = (ms: Array<{ id?: string }>) => ms.map(m => m.id as string);

const rows: unknown[] = [];

// ---- filterMessagesByHistoryAccess ---------------------------------------
{
  type HMsg = { id: string; createdAt?: string };
  const cases: Array<{ id: string; msgs: HMsg[]; hasHistoryAccess: boolean; joinCreatedAt: string }> = [
    {
      id: 'full-access-keeps-all',
      msgs: [{ id: 'm1', createdAt: '2026-01-01T00:00:00.000Z' }, { id: 'm2', createdAt: '2026-02-01T00:00:00.000Z' }],
      hasHistoryAccess: true,
      joinCreatedAt: '2026-03-01T00:00:00.000Z',
    },
    {
      id: 'after-join-only',
      msgs: [
        { id: 'before', createdAt: '2026-01-01T00:00:00.000Z' },
        { id: 'at-join', createdAt: '2026-02-01T00:00:00.000Z' },
        { id: 'after', createdAt: '2026-03-01T00:00:00.000Z' },
      ],
      hasHistoryAccess: false,
      joinCreatedAt: '2026-02-01T00:00:00.000Z',
    },
    {
      id: 'no-createdAt-kept',
      msgs: [{ id: 'nodate' }, { id: 'old', createdAt: '2025-01-01T00:00:00.000Z' }],
      hasHistoryAccess: false,
      joinCreatedAt: '2026-01-01T00:00:00.000Z',
    },
  ];
  for (const c of cases) {
    const participant = asPart({ hasHistoryAccess: c.hasHistoryAccess, createdAt: c.joinCreatedAt });
    const out = filterMessagesByHistoryAccess(asMsgs(c.msgs), participant);
    rows.push({
      kind: 'history',
      id: c.id,
      msgs: c.msgs.map(m => ({ id: m.id, createdAtMs: m.createdAt ? new Date(m.createdAt).getTime() : null })),
      hasHistoryAccess: c.hasHistoryAccess,
      joinMs: new Date(c.joinCreatedAt).getTime(),
      out: idsOf(out as Array<{ id?: string }>),
    });
  }
}

// ---- computePresenceWindowsForParticipant --------------------------------
{
  const hev = (id: string, at: string, participantId: string, toStatus?: string) =>
    ({ id, createdAt: at, hostEvent: toStatus === undefined ? { introducedCharacterIds: ['x'] } : { participantId, toStatus } });
  const cases: Array<{ id: string; msgs: unknown[]; participantId: string; participantCreatedAt: string }> = [
    { id: 'no-events-open-window', msgs: [], participantId: 'P', participantCreatedAt: '2026-01-01T00:00:00.000Z' },
    {
      id: 'active-then-absent',
      msgs: [hev('e1', '2026-01-02T00:00:00.000Z', 'P', 'active'), hev('e2', '2026-01-05T00:00:00.000Z', 'P', 'absent')],
      participantId: 'P', participantCreatedAt: '2026-01-01T00:00:00.000Z',
    },
    {
      id: 'reopen-second-open',
      msgs: [
        hev('e1', '2026-01-02T00:00:00.000Z', 'P', 'active'),
        hev('e2', '2026-01-05T00:00:00.000Z', 'P', 'absent'),
        hev('e3', '2026-01-08T00:00:00.000Z', 'P', 'silent'),
      ],
      participantId: 'P', participantCreatedAt: '2026-01-01T00:00:00.000Z',
    },
    {
      id: 'other-participant-ignored',
      msgs: [hev('e1', '2026-01-02T00:00:00.000Z', 'OTHER', 'absent')],
      participantId: 'P', participantCreatedAt: '2026-01-01T00:00:00.000Z',
    },
    {
      id: 'introduction-no-tostatus-ignored',
      msgs: [hev('e1', '2026-01-02T00:00:00.000Z', 'P', undefined)],
      participantId: 'P', participantCreatedAt: '2026-01-01T00:00:00.000Z',
    },
    {
      id: 'out-of-order-sorted',
      msgs: [hev('e2', '2026-01-05T00:00:00.000Z', 'P', 'absent'), hev('e1', '2026-01-02T00:00:00.000Z', 'P', 'active')],
      participantId: 'P', participantCreatedAt: '2026-01-01T00:00:00.000Z',
    },
  ];
  for (const c of cases) {
    const participant = asPart({ id: c.participantId, createdAt: c.participantCreatedAt });
    const out = computePresenceWindowsForParticipant(asMsgs(c.msgs), participant);
    rows.push({ kind: 'presence', id: c.id, msgs: c.msgs, participantId: c.participantId, participantCreatedAt: c.participantCreatedAt, out });
  }
}

// ---- filterMessagesByPresenceWindows -------------------------------------
{
  type PMsg = { id: string; createdAt?: string };
  const cases: Array<{ id: string; msgs: PMsg[]; windows: PresenceWindow[] }> = [
    {
      id: 'empty-windows-drop-all',
      msgs: [{ id: 'm1', createdAt: '2026-01-02T00:00:00.000Z' }],
      windows: [],
    },
    {
      id: 'closed-window-bounds',
      msgs: [
        { id: 'before', createdAt: '2026-01-01T00:00:00.000Z' },
        { id: 'at-from', createdAt: '2026-01-02T00:00:00.000Z' },
        { id: 'inside', createdAt: '2026-01-03T00:00:00.000Z' },
        { id: 'at-to', createdAt: '2026-01-05T00:00:00.000Z' },
        { id: 'nodate' },
      ],
      windows: [{ from: '2026-01-02T00:00:00.000Z', to: '2026-01-05T00:00:00.000Z' }],
    },
    {
      id: 'open-window-keeps-tail',
      msgs: [
        { id: 'before', createdAt: '2026-01-01T00:00:00.000Z' },
        { id: 'after', createdAt: '2026-09-01T00:00:00.000Z' },
      ],
      windows: [{ from: '2026-02-01T00:00:00.000Z', to: null }],
    },
    {
      id: 'two-windows',
      msgs: [
        { id: 'w1', createdAt: '2026-01-03T00:00:00.000Z' },
        { id: 'gap', createdAt: '2026-01-06T00:00:00.000Z' },
        { id: 'w2', createdAt: '2026-01-09T00:00:00.000Z' },
      ],
      windows: [
        { from: '2026-01-02T00:00:00.000Z', to: '2026-01-05T00:00:00.000Z' },
        { from: '2026-01-08T00:00:00.000Z', to: null },
      ],
    },
  ];
  for (const c of cases) {
    const out = filterMessagesByPresenceWindows(asMsgs(c.msgs), c.windows);
    rows.push({ kind: 'presenceFilter', id: c.id, msgs: c.msgs, windows: c.windows, out: idsOf(out as Array<{ id?: string }>) });
  }
}

// ---- filterWhisperMessages -----------------------------------------------
{
  type WMsg = { id: string; participantId?: string | null; targetParticipantIds?: string[] | null };
  const cases: Array<{ id: string; msgs: WMsg[]; respondingId: string }> = [
    {
      id: 'visibility',
      msgs: [
        { id: 'public', participantId: 'A' },
        { id: 'empty-targets', participantId: 'A', targetParticipantIds: [] },
        { id: 'to-others', participantId: 'A', targetParticipantIds: ['B'] },
        { id: 'sender-sees', participantId: 'R', targetParticipantIds: ['B'] },
        { id: 'target-sees', participantId: 'A', targetParticipantIds: ['R', 'B'] },
      ],
      respondingId: 'R',
    },
  ];
  for (const c of cases) {
    const out = filterWhisperMessages(asMsgs(c.msgs), c.respondingId);
    rows.push({ kind: 'whisper', id: c.id, msgs: c.msgs, respondingId: c.respondingId, out: idsOf(out as Array<{ id?: string }>) });
  }
}

// ---- getParticipantName --------------------------------------------------
{
  const parts = [
    { id: 'pc', type: 'CHARACTER', characterId: 'c-known', controlledBy: 'llm', status: 'active' },
    { id: 'pc-unknown', type: 'CHARACTER', characterId: 'c-missing', controlledBy: 'llm', status: 'active' },
    { id: 'pc-empty-cid', type: 'CHARACTER', characterId: '', controlledBy: 'llm', status: 'active' },
    { id: 'pc-noname', type: 'CHARACTER', characterId: 'c-empty', controlledBy: 'llm', status: 'active' },
    { id: 'p-other', type: 'OTHER', characterId: 'c-known', controlledBy: 'llm', status: 'active' },
  ];
  const chars = { 'c-known': 'Friday', 'c-empty': '' };
  const cases: Array<{ id: string; participantId: string | null }> = [
    { id: 'null-id', participantId: null },
    { id: 'missing-participant', participantId: 'nope' },
    { id: 'known-character', participantId: 'pc' },
    { id: 'character-not-in-map', participantId: 'pc-unknown' },
    { id: 'empty-characterId', participantId: 'pc-empty-cid' },
    { id: 'empty-name', participantId: 'pc-noname' },
    { id: 'non-character-type', participantId: 'p-other' },
  ];
  for (const c of cases) {
    const out = getParticipantName(c.participantId, asChars(chars), asParts(parts));
    rows.push({ kind: 'name', id: c.id, participantId: c.participantId, characters: chars, participants: parts, out: out ?? null });
  }
}

// ---- attributeMessagesForCharacter ---------------------------------------
{
  const parts = [
    { id: 'R', type: 'CHARACTER', characterId: 'c-r', controlledBy: 'llm', status: 'active' },
    { id: 'O', type: 'CHARACTER', characterId: 'c-o', controlledBy: 'llm', status: 'active' },
  ];
  const chars = { 'c-r': 'Responder', 'c-o': 'Other' };
  const msgs = [
    { id: 'self', role: 'ASSISTANT', content: 'mine', participantId: 'R', thoughtSignature: 'sig1' },
    { id: 'other', role: 'ASSISTANT', content: 'theirs', participantId: 'O', thoughtSignature: null },
    { id: 'usermsg', role: 'USER', content: 'hi', participantId: null },
    { id: 'empty-pid', role: 'USER', content: 'blank', participantId: '' },
  ];
  const out = attributeMessagesForCharacter(asMsgs(msgs), 'R', asChars(chars), asParts(parts));
  rows.push({ kind: 'attribute', id: 'responder-view', msgs, respondingId: 'R', characters: chars, participants: parts, out });
}

// ---- findUserParticipantName ---------------------------------------------
{
  const chars = { 'c-user': 'Bertie', 'c-user2': 'Jeeves', 'c-blank': '' };
  const cases: Array<{ id: string; participants: unknown[]; activeTyping: string | null }> = [
    {
      id: 'active-typing-selected',
      participants: [
        { id: 'u1', type: 'CHARACTER', characterId: 'c-user', controlledBy: 'user', status: 'active' },
        { id: 'u2', type: 'CHARACTER', characterId: 'c-user2', controlledBy: 'user', status: 'active' },
      ],
      activeTyping: 'u2',
    },
    {
      id: 'active-typing-invalid-falls-back',
      participants: [
        { id: 'u1', type: 'CHARACTER', characterId: 'c-user', controlledBy: 'user', status: 'active' },
        { id: 'u2', type: 'CHARACTER', characterId: 'c-user2', controlledBy: 'user', status: 'absent' },
      ],
      activeTyping: 'u2',
    },
    {
      id: 'no-active-typing-first-user',
      participants: [
        { id: 'llm', type: 'CHARACTER', characterId: 'c-user2', controlledBy: 'llm', status: 'active' },
        { id: 'u1', type: 'CHARACTER', characterId: 'c-user', controlledBy: 'user', status: 'active' },
      ],
      activeTyping: null,
    },
    {
      id: 'no-user-character',
      participants: [
        { id: 'llm', type: 'CHARACTER', characterId: 'c-user2', controlledBy: 'llm', status: 'active' },
      ],
      activeTyping: null,
    },
    {
      id: 'blank-name-falls-through',
      participants: [
        { id: 'u1', type: 'CHARACTER', characterId: 'c-blank', controlledBy: 'user', status: 'active' },
      ],
      activeTyping: null,
    },
  ];
  for (const c of cases) {
    const out = findUserParticipantName(asParts(c.participants), asChars(chars), c.activeTyping);
    rows.push({ kind: 'userName', id: c.id, participants: c.participants, characters: chars, activeTyping: c.activeTyping, out: out ?? null });
  }
}

for (const r of rows) process.stdout.write(JSON.stringify(r) + '\n');

File: crates/quilltap-core/src/message_attribution.rs
//! Port of the pure history-access / presence / attribution helpers in v4's
//! lib/chat/context/message-attribution.ts. These shape the conversation as a
//! particular responding character sees it: history-access gating, presence
//! windows, whisper visibility, and per-character role/name attribution.
//!
//! The one impurity in v4 is `new Date(createdAt).getTime()` inside the
//! history-access filter; here it is injected as pre-parsed epoch millis (the
//! same seam used elsewhere in the port). The presence-window filters compare
//! ISO strings lexically exactly as v4 does — lex order matches chronological
//! order for the uniform `…Z` timestamps the data uses.

use std::collections::HashMap;

use crate::chat_predicates::{is_participant_present, ParticipantStatus};

/// A participant as the attribution helpers read it. Not every field is used by
/// every function (history-access reads `has_history_access`/join time;
/// name attribution reads `participant_type`/`character_id`; etc.).
#[derive(Clone, Debug)]
pub struct AttributionParticipant {
    pub id: String,
    pub participant_type: String,
    pub character_id: Option<String>,
    pub controlled_by: String,
    pub status: ParticipantStatus,
}

/// A Host status announcement payload. Presence tracking filters on the
/// `(participant_id, to_status)` shape; off-scene introductions (only
/// `introducedCharacterIds` set) carry `to_status: None` and are ignored.
#[derive(Clone, Debug)]
pub struct HostEvent {
    pub participant_id: Option<String>,
    pub to_status: Option<ParticipantStatus>,
}

/// A message in multi-character context building. `created_at` is the ISO
/// string (used lexically by the presence filters).
#[derive(Clone, Debug)]
pub struct AttributionMessage {
    pub id: Option<String>,
    pub role: String,
    pub content: String,
    pub participant_id: Option<String>,
    pub thought_signature: Option<String>,
    pub created_at: Option<String>,
    pub target_participant_ids: Option<Vec<String>>,
    pub host_event: Option<HostEvent>,
}

/// A message reduced to what the history-access numeric filter needs: its
/// creation instant in epoch millis (None when `createdAt` was absent — such a
/// message is always kept). The ISO→ms parse is the injected date seam.
#[derive(Clone, Debug)]
pub struct HistoryMessage {
    pub created_at_ms: Option<f64>,
}

/// A presence window: an interval `[from, to)` during which the participant was
/// 'active' or 'silent'. `to` is `None` on the trailing open window. Times are
/// ISO strings; lex order matches chronological order.
#[derive(Clone, Debug, PartialEq, Eq)]
pub struct PresenceWindow {
    pub from: String,
    pub to: Option<String>,
}

/// The result of attributing a message to a responding character's viewpoint.
#[derive(Clone, Debug, PartialEq, Eq)]
pub struct MultiCharacterMessage {
    pub role: &'static str,
    pub content: String,
    pub id: Option<String>,
    pub name: Option<String>,
    pub participant_id: Option<String>,
    pub thought_signature: Option<String>,
}

/// Filter messages by a participant's history access. With full access, every
/// message is returned. Otherwise only messages at or after the participant's
/// join time survive; a message with no `createdAt` (`None` ms) is kept
/// (shouldn't happen). Returns a keep-mask aligned to `messages`.
///
/// `join_ms` is `new Date(participant.createdAt).getTime()` and each message's
/// `created_at_ms` is the same parse — both injected at the boundary.
pub fn filter_messages_by_history_access(
    messages: &[HistoryMessage],
    has_history_access: bool,
    join_ms: f64,
) -> Vec<bool> {
    if has_history_access {
        return vec![true; messages.len()];
    }
    messages
        .iter()
        .map(|m| match m.created_at_ms {
            None => true,
            Some(t) => t >= join_ms,
        })
        .collect()
}

/// Compute the presence windows for a participant by walking Host status
/// announcements (`hostEvent.participantId === participant.id`) in chronological
/// order. An interval is open while the participant is `active`/`silent` and
/// closed by a transition to `absent`/`removed`; the trailing open window has
/// `to: None`. With no Host events, the participant is assumed present from
/// `participant_created_at` onward (one open window).
pub fn compute_presence_windows_for_participant(
    messages: &[AttributionMessage],
    participant_id: &str,
    participant_created_at: &str,
) -> Vec<PresenceWindow> {
    let mut events: Vec<(String, ParticipantStatus)> = messages
        .iter()
        .filter_map(|m| {
            let he = m.host_event.as_ref()?;
            let to_status = he.to_status?;
            let created_at = m.created_at.as_ref()?;
            if he.participant_id.as_deref() == Some(participant_id) {
                Some((created_at.clone(), to_status))
            } else {
                None
            }
        })
        .collect();

    // Stable sort by ISO instant, matching v4's comparator.
    events.sort_by(|a, b| a.0.cmp(&b.0));

    if events.is_empty() {
        return vec![PresenceWindow {
            from: participant_created_at.to_string(),
            to: None,
        }];
    }

    let mut windows: Vec<PresenceWindow> = Vec::new();
    let mut open_from: Option<String> = None;
    for (at, to_status) in events {
        let present = is_participant_present(to_status);
        if present {
            if open_from.is_none() {
                open_from = Some(at);
            }
        } else if let Some(from) = open_from.take() {
            windows.push(PresenceWindow { from, to: Some(at) });
        }
    }
    if let Some(from) = open_from {
        windows.push(PresenceWindow { from, to: None });
    }

    windows
}

/// Keep messages whose `createdAt` falls inside one of the presence windows.
/// `from` inclusive, `to` exclusive; a `None` `to` means still open. Messages
/// without `createdAt` are dropped. Empty window set drops everything. Returns
/// a keep-mask aligned to `messages`.
pub fn filter_messages_by_presence_windows(
    messages: &[AttributionMessage],
    windows: &[PresenceWindow],
) -> Vec<bool> {
    if windows.is_empty() {
        return vec![false; messages.len()];
    }
    messages
        .iter()
        .map(|m| match m.created_at.as_deref() {
            None => false,
            Some(t) => windows
                .iter()
                .any(|w| t >= w.from.as_str() && w.to.as_deref().is_none_or(|to| t < to)),
        })
        .collect()
}

/// Keep whisper messages visible to the responding participant. A public
/// message (no `targetParticipantIds`) is always visible; a whisper is visible
/// to its sender and its targets only. Returns a keep-mask aligned to
/// `messages`.
pub fn filter_whisper_messages(
    messages: &[AttributionMessage],
    responding_participant_id: &str,
) -> Vec<bool> {
    messages
        .iter()
        .map(|m| {
            match &m.target_participant_ids {
                // Public message — always include.
                None => true,
                Some(targets) if targets.is_empty() => true,
                Some(targets) => {
                    // Sender sees their own whispers; targets see whispers at them.
                    m.participant_id.as_deref() == Some(responding_participant_id)
                        || targets.iter().any(|t| t == responding_participant_id)
                }
            }
        })
        .collect()
}

/// Resolve a participant's display name for attribution. CHARACTER participants
/// (LLM or user-controlled) resolve via their `characterId` into the character
/// map. Returns `None` when the id is absent, the participant is missing, or it
/// isn't a CHARACTER with a known character.
pub fn get_participant_name(
    participant_id: Option<&str>,
    participant_characters: &HashMap<String, String>,
    all_participants: &[AttributionParticipant],
) -> Option<String> {
    let participant_id = participant_id?;
    let participant = all_participants.iter().find(|p| p.id == participant_id)?;

    // v4 guards on `participant.characterId` truthiness — empty string is falsy.
    if participant.participant_type == "CHARACTER" {
        if let Some(cid) = participant.character_id.as_deref() {
            if !cid.is_empty() {
                return participant_characters.get(cid).cloned();
            }
        }
    }
    None
}

/// Attribute messages for a responding character's viewpoint: messages from the
/// responder become `assistant`; everything else becomes `user`, carrying the
/// sender's name for disambiguation.
pub fn attribute_messages_for_character(
    messages: &[AttributionMessage],
    responding_participant_id: &str,
    participant_characters: &HashMap<String, String>,
    all_participants: &[AttributionParticipant],
) -> Vec<MultiCharacterMessage> {
    messages
        .iter()
        .map(|msg| {
            let participant_name = get_participant_name(
                msg.participant_id.as_deref(),
                participant_characters,
                all_participants,
            );

            // assistant iff the message is from the responding character; the
            // other-character and USER branches both yield `user` in v4.
            let role = if msg.participant_id.as_deref() == Some(responding_participant_id) {
                "assistant"
            } else {
                "user"
            };

            // `msg.participantId || undefined` — empty string drops to None.
            let participant_id = match msg.participant_id.as_deref() {
                Some(p) if !p.is_empty() => Some(p.to_string()),
                _ => None,
            };

            MultiCharacterMessage {
                role,
                content: msg.content.clone(),
                id: msg.id.clone(),
                name: participant_name,
                participant_id,
                thought_signature: msg.thought_signature.clone(),
            }
        })
        .collect()
}

/// Find the user participant's display name for attribution. Prefers the
/// actively-selected "Speaking As" participant when it is a present,
/// user-controlled CHARACTER; otherwise the first such participant. A character
/// with no (truthy) name yields `None`.
pub fn find_user_participant_name(
    all_participants: &[AttributionParticipant],
    participant_characters: &HashMap<String, String>,
    active_typing_participant_id: Option<&str>,
) -> Option<String> {
    let is_user_character = |p: &AttributionParticipant| {
        p.participant_type == "CHARACTER"
            && p.controlled_by == "user"
            && is_participant_present(p.status)
            && p.character_id.as_deref().is_some_and(|c| !c.is_empty())
    };

    let selected = active_typing_participant_id.and_then(|atid| {
        all_participants
            .iter()
            .find(|p| p.id == atid && is_user_character(p))
    });
    let user_character_participant =
        selected.or_else(|| all_participants.iter().find(|p| is_user_character(p)));

    if let Some(p) = user_character_participant {
        if let Some(cid) = p.character_id.as_deref() {
            if let Some(name) = participant_characters.get(cid) {
                if !name.is_empty() {
                    return Some(name.clone());
                }
            }
        }
    }
    None
}
